pkg/server/delta/v3/server.go

```diff
@@ -44,6 +54,211 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+			// reset it on subsequent requests that omit it.
+			if req.Node != nil {
+				node = req.Node
+				nonce = req.GetResponseNonce()
+			} else {
+				req.Node = node
+
+				// If we have no nonce, i.e. this is the first request on a delta stream, set one
+				nonce = strconv.FormatInt(streamNonce, 10)
+			}
+
+			// type URL is required for ADS but is implicit for any other xDS stream
+			if defaultTypeURL == resource.AnyType {
+				if req.TypeUrl == "" {
+					return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+				}
+			} else if req.TypeUrl == "" {
+				req.TypeUrl = defaultTypeURL
+			}
+
+			typeURL := req.GetTypeUrl()
+			state, ok := watches.deltaStreamStates[typeURL]
+			if !ok {
+				// Initialize the state if we haven't already.
+				state = stream.NewStreamState()
+				// Since there was no previous state, we know we're handling the first request of this type on this stream
+				for r, v := range req.InitialResourceVersions {
+					state.ResourceVersions[r] = v
+				}
+			}
+			versionMap := state.GetResourceVersions()
+
+			s.subscribe(req.GetResourceNamesSubscribe(), versionMap)
+			s.unsubscribe(req.GetResourceNamesUnsubscribe(), versionMap)
+
+			// cancel existing watch to (re-)request a newer version
+			watch, ok := watches.deltaResponses[typeURL]
+			// we verify nonce only if nonce is not initialized
+			if !ok || watch.nonce == nonce {
+				// We must signal goroutine termination to prevent a race between the cancel closing the watch
+				// and the producer closing the watch.`	`
+				if terminate, exists := watches.deltaTerminations[typeURL]; exists {
+					close(terminate)
+				}
+				if watch.cancel != nil {
+					watch.cancel()
+				}
+				watch.responses, watch.cancel = s.cache.CreateDeltaWatch(req, versionMap)
+
+				// Go does not allow for selecting over a dynamic set of channels
+				// so we introduce a termination chan to handle cancelling any watches.
+				terminate := make(chan struct{})
+				watches.deltaTerminations[typeURL] = terminate
+				go func() {
+					select {
+					case resp, more := <-watch.responses:
+						if more {
+							watches.deltaMuxedResponses <- resp
+							state.SetResourceVersions(resp.GetNextVersionMap())
+						} else {
+							// Check again if the watch is cancelled.
+							select {
+							case <-terminate: // do nothing
+							default:
+								// We cannot close the responses channel since it can be closed twice.
+								// Instead we send a fake error response.
+								watches.deltaMuxedResponses <- deltaErrorResponse
+							}
+						}
+						break
+					case <-terminate:
+						break
+					}
+				}()
+			}
+		}
+	}
+}
+
 func (s *server) DeltaStreamHandler(str stream.DeltaStream, typeURL string) error {
-	return nil
+	// a channel for receiving incoming delta requests
+	reqCh := make(chan *discovery.DeltaDiscoveryRequest)
+
+	// we need to concurrently handle incoming requests since we kick off processDelta as a return
```

Comment:
I think we do this so we can `select` between requests and watch triggers, maybe rephrase in terms of that?

Reply:
Ah yes i'll make this change

Reply:
@snowp I re-read this after I got a bit confused, wouldn't that only be the case if we were actually calling `processDelta` inside of the select? I think the the point of the select is to see if we need to close down the request channel if the stream closes itself. Maybe I'm misinterpreting that logic though